platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +90,196 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
+          ServerErrorException.class);
+    }
+
+    // buy_amount has a proper number of significant decimals
+    if (!hasProperSignificantDecimals(
+        Objects.requireNonNull(rate).getBuyAmount(), buyAsset.getSignificantDecimals())) {
+      logErrorAndThrow(
+          format(
+              "'rate.buy_amount' (%s) has incorrect number of significant decimals (expected: %d) in the GET /rate response",
+              rate.getBuyAmount(), buyAsset.getSignificantDecimals()),
+          ServerErrorException.class);
+    }
+
+    FeeDetails fee = rate.getFee();
+    // if fee is set, check the following
+    if (fee != null) {
+      // fee.total is present and is a positive number
+      if (!isPositiveNumber(fee.getTotal())) {
+        logErrorAndThrow(
+            "'rate.fee.total' is missing or not a positive number in the GET /rate response",
+            ServerErrorException.class);
+      }
+      // fee.asset is a valid asset
+      AssetInfo feeAsset = assetService.getAssetByName(fee.getAsset());
+      if (fee.getAsset() == null || feeAsset == null) {
+        logErrorAndThrow(
+            "'rate.fee.asset' is missing or not a valid asset in the GET /rate response",
+            ServerErrorException.class);
+      }
+
+      if (!isPositiveNumber(Objects.requireNonNull(feeAsset.getSignificantDecimals()).toString())) {
+        logErrorAndThrow(
+            format(
+                "The rate fee.asset (%s) does not have proper significant decimals defined in the assets configuration.",
+                feeAsset.getSignificantDecimals()),
+            ServerErrorException.class);
+      }
+
+      // fee.total has a proper number of significant decimals
+      if (!hasProperSignificantDecimals(fee.getTotal(), feeAsset.getSignificantDecimals())) {
+        logErrorAndThrow(
+            format(
+                "'rate.fee.total' (%s) has incorrect number of significant decimals (expected: %d) in the GET /rate response",
+                fee.getTotal(), feeAsset.getSignificantDecimals()),
+            ServerErrorException.class);
+      }
+
+      if (fee.getAsset().equals(request.getSellAsset())) {
+        // when fee is in sell_asset,
+        // check that sell_amount ~= price * buy_amount + (fee ?: 0)
+        BigDecimal expected =
+            new BigDecimal(rate.getPrice())
+                .multiply(new BigDecimal(rate.getBuyAmount()))
+                .add(new BigDecimal(fee.getTotal()));
+
+        // Since we don't know how the anchor rounds the amounts, we need to check the equality in
+        // all allowed rounding modes
+        if (!withinRoundingError(
+            new BigDecimal(rate.getSellAmount()), expected, sellAsset.getSignificantDecimals())) {
+          logErrorAndThrow(
+              format(
+                  "'rate.sell_amount' (%s) is not within rounding error of the expected (%s) ('price * buy_amount + (fee?:0)') in the GET /rate response",
```

Comment:
fee here is non-zero, so just `price * buy_amount + fee`

Reply:
Fixed.